src/Futhark/CodeGen/ImpGen/GPU/SegScan/SinglePass.hs

```diff
@@ -314,7 +316,11 @@ compileSegScan pat lvl space scan_op map_kbody = do
                 (untyped (1 :: Imp.TExp Int32))
           sComment "Set dynamic id and reset status flag for this block" $ do
             copyDWIMFix sharedId [0] (tvSize dyn_id) []
-            copyDWIMFix statusFlags [tvExp dyn_id] (intConst Int8 statusX) []
+            -- TODO: find out why this flag initialization (rather than
+            -- initializing using sReplicate outside the current kernel) is
+            -- causing problems. it was never a problem with any of our
+            -- prototypes.
+            -- copyDWIMFix statusFlags [tvExp dyn_id] (intConst Int8 statusX) []
```

Comment:
There is no guarantee any of the previous blocks have had a chance to set their status flags by the time other blocks start reading them, is there?  My intuition is certainly that it must be *fully* done before the kernel starts.

Reply:
Well, this is embarassing. Yes, your intuition is of course correct. It seems that I (and, in my cowardly defense, the other prototype on which we based experiments) falsely applied the same logic onto the status flags as for the global ID counter, which _can_ be initialized once and efficiently reset inside the kernel.

The race is very unlikely, since it requires a thread group to both fetch its dynamic ID `j`, set `status_flags[j] = flag_X`, compute an intra-group scan, publish its aggregate to global mem, and start its lookback phase -- all in the time it takes between some other group fetching its dynamic ID `i < j` (this is guaranteed to happen before group `j` fetches theirs) and setting `status_flags[i] = flag_X` in the very next line of code. But yes, obviously possible!